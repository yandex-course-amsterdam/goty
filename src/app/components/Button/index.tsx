import React, { ReactElement } from 'react'
import cn from 'classnames'

import style from './style.css'

type ButtonProps = {
  type: 'button' | 'submit'
  buttonText: string
  className?: string
  handleClick?: () => void
}

export const Button = ({ type, buttonText, className, handleClick }: ButtonProps): ReactElement => {
  return (
    <button
      onClick={handleClick}
      className={cn(style.button, className)}
      type={type === 'submit' ? 'submit' : 'button'}
      disabled
    >
      {buttonText}
    </button>
  )
}
