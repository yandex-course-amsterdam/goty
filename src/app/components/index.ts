export { Sidebar } from './Sidebar'
export { Avatar } from './Avatar'
export { Navigation } from './Navigation'
export { Main } from './Main'
export { Title } from './Title'
export { Description } from './Description'
export { Input } from './Input'
export { Textarea } from './Textarea'
export { Label } from './Label'
export { Error } from './Error'
export { ErrorBoundary } from './ErrorBoundary'
export { OfflineBar } from './OfflineBar'
export { Button } from './Button'
export { SubNavigation } from './SubNavigation'
export { Canvas } from './Canvas'
export { UserScore } from './UserScore'
export { Comment } from './Comment'
export { Comments } from './Comments'
export { Likes } from './Likes'
export { NewsList } from './NewsList'
export { SignUpForm } from './Forms/SignUpForm'
export { SignInForm } from './Forms/SignInForm'
export { FeedbackForm } from './Forms/FeedbackForm'
export { DetailsForm } from './Forms/DetailsForm'
export { PasswordForm } from './Forms/PasswordForm'
export { AvatarForm } from './Forms/AvatarForm'
export { Loader } from './Loader'
export { PrivateRoute } from './PrivateRoute'
export { ExitButton } from './ExitButton'
export { Authorization } from './Authorization'
export { Theme } from './Theme'
export { App } from './App'
