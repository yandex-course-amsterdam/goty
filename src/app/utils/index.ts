export { sanitize } from './sanitize'
export { displayResponseText } from './displayResponseText'
export { startServiceWorker } from './startServiceWorker'
