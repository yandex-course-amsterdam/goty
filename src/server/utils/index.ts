export { createToken } from './createToken'
export { verifyToken } from './verifyToken'
