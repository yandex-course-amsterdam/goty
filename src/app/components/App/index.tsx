import React, { FC, useCallback, useEffect, useState } from 'react'
import { useSelector, useDispatch } from 'react-redux'

import { Switch, Route } from 'react-router-dom'
import { CSSTransition } from 'react-transition-group'

import {
  SignUpView,
  SignInView,
  ProfileView,
  GameView,
  ScoreView,
  NotFoundView
} from 'app/views'
import { OfflineBar, PrivateRoute, Authorization } from 'app/components'

import { postResult, getUserTheme } from 'app/api/Api'
import { fetchUserInfo } from 'app/actions'

import { StoreState } from 'app/reducers'
import { route } from 'app/enums'
import {
  getScore,
  removeScore,
  isServer,
  setUserTheme,
  storeUserTheme
} from 'app/utils'

import 'normalize.css'
import 'assets/main.css'
import '../../../fonts/fonts.css'

// startServiceWorker()

export const App: FC = (): JSX.Element => {
  const [isOffline, setIsOffline] = useState(isServer || !navigator.onLine)

  const dispatch = useDispatch()

  const loginStatus = useSelector(
    (state: StoreState) => state.loginStatus.status
  )

  const handleOfflineCallback = useCallback(() => setIsOffline(true), [])
  const handleOnlineCallback = useCallback(() => {
    setIsOffline(false)

    const storedScore = getScore()

    if (storedScore) {
      postResult(storedScore)
      removeScore()
    }
  }, [])

  useEffect(() => {
    const fetchUser = async () => {
      try {
        const data = await dispatch(fetchUserInfo())
        // Очередной странный тайпинг, учитывая возвращаемое по факту значение
        // Сейчас уже нет времени разбираться, буду потом сильнее вникать в ts
        // @ts-ignore
        const { data: theme } = await getUserTheme(data.id)
        setUserTheme(theme)
        storeUserTheme(theme)
      } catch (error) {
        console.warn(error)
      }
    }

    if (loginStatus) {
      fetchUser()
    }

    if (!isServer) {
      window.addEventListener('offline', handleOfflineCallback)
      window.addEventListener('online', handleOnlineCallback)
    }

    return () => {
      if (!isServer) {
        window.removeEventListener('offline', handleOfflineCallback)
        window.removeEventListener('online', handleOnlineCallback)
      }
    }
  }, [])

  return (
    <>
      <Switch>
        <Route exact path={route.auth}>
          <Authorization />
        </Route>
        <Route path={route.signUp}>
          <SignUpView />
        </Route>
        <Route path={route.signIn}>
          <SignInView />
        </Route>
        <PrivateRoute path={route.profile}>
          <ProfileView />
        </PrivateRoute>
        <PrivateRoute path={route.game}>
          <GameView />
        </PrivateRoute>
        <PrivateRoute path={route.score}>
          <ScoreView />
        </PrivateRoute>
        <Route path="*">
          <NotFoundView />
        </Route>
      </Switch>

      <CSSTransition
        in={isOffline}
        appear={isOffline}
        timeout={0}
        classNames="bar"
      >
        <OfflineBar />
      </CSSTransition>
    </>
  )
}
