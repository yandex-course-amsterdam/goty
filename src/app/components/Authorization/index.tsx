import React, { FC, useEffect, useState } from 'react'
import { useDispatch } from 'react-redux'
import { useLocation, Redirect } from 'react-router-dom'

import { authWithYandexOauth } from 'app/api/Api'
import { fetchUserInfo, setOAuthStatus } from 'app/actions'
import { LoaderView } from 'app/views'
import { route } from 'app/enums'

const useQuery = () => new URLSearchParams(useLocation().search)

export const Authorization: FC = (): JSX.Element => {
  const [isLoading, setIsLoading] = useState(true)
  const dispatch = useDispatch()

  const query = useQuery()
  const codeString = query.get('code')

  const getUserData = async () => {
    try {
      await dispatch(fetchUserInfo())
      setIsLoading(!isLoading)
    } catch (error) {
      console.error(error)
      setIsLoading(!isLoading)
    }
  }

  const signInWithOAuth = async () => {
    try {
      await authWithYandexOauth(codeString)
      await dispatch(fetchUserInfo())
      dispatch(setOAuthStatus(true))
      setIsLoading(!isLoading)
    } catch (error) {
      console.log(error)
      setIsLoading(!isLoading)
    }
  }

  useEffect(() => {
    if (codeString) {
      signInWithOAuth()
    } else {
      getUserData()
    }
  }, [])

  return isLoading ? <LoaderView /> : <Redirect to={route.game} />
}
