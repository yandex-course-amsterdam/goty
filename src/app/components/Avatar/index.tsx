import React, { FC } from 'react'
import { useSelector } from 'react-redux'
import cn from 'classnames'

import { StoreState } from 'app/reducers'

import defaultAvatar from 'images/avatar.png'

import style from './style.css'

interface IProps {
  className?: string
}

export const Avatar: FC<IProps> = ({ className }): JSX.Element => {
  const userInfo = useSelector((state: StoreState) => state.userInfo)

  const userAvatar = userInfo.avatar
    ? `https://ya-praktikum.tech/${userInfo.avatar}`
    : defaultAvatar

  const avatarName = userInfo.display_name || 'Top game'

  return (
    <div className={cn(style.avatar, className)}>
      <img className={style.image} src={userAvatar} alt={avatarName} />
      <h1 className={style.title}>{avatarName}</h1>
    </div>
  )
}
