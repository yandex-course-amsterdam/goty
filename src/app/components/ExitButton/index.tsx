import React, { useState, FC } from 'react'
import { Redirect } from 'react-router-dom'
import cn from 'classnames'

import { authApi } from 'app/api'
import { route } from 'app/enums'

import style from './style.css'

interface IProps {
  children: React.ReactNode
  className?: string
}

export const ExitButton: FC<IProps> = ({
  children,
  className
}): JSX.Element => {
  const [isLoggedIn, setIsLoggedIn] = useState(true)

  const handleExit = async () => {
    try {
      const res = await authApi.logout()

      if (res.status === 200) {
        setIsLoggedIn(false)
      }
    } catch (error) {
      console.log(error)
    }
  }

  return isLoggedIn ? (
    <button
      type="button"
      className={cn(style.button, className)}
      onClick={handleExit}
    >
      {children}
    </button>
  ) : (
    <Redirect to={route.signIn} />
  )
}
