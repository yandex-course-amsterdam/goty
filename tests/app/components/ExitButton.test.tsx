import React from 'react'
import * as renderer from 'react-test-renderer'
import { createMemoryHistory } from 'history'

import { ExitButton } from 'app/components/ExitButton'

import { renderWithRouter } from '../../utils'

const history = createMemoryHistory()

// eslint-disable-next-line
// @ts-ignore
jest.spyOn(React, 'useState').mockImplementationOnce(() => [false])

describe('Exit button component', () => {
  test('Render redirect if user is not logged in', () => {
    const component = renderer
      .create(renderWithRouter(<ExitButton>Oops</ExitButton>, history))
      .toJSON()
    // Snapshot equals to null when there are no parent container for element rendered directly inside Router component
    // Сompare to Navigation component snapshots for example
    // ¯\_(ツ)_/¯
    expect(component).toMatchSnapshot()
  })

  test('Render default markup if user is logged in', () => {
    const component = renderer.create(<ExitButton>Oops</ExitButton>).toJSON()
    expect(component).toMatchSnapshot()
  })

  test('Use className prop', () => {
    const component = renderer
      .create(<ExitButton className="someClass">Oops</ExitButton>)
      .toJSON()
    expect(component).toMatchSnapshot()
  })
})
