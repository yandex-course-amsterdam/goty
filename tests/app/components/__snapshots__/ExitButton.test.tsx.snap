// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Exit button component Render default markup if user is logged in 1`] = `
<button
  className=""
  onClick={[Function]}
  type="button"
>
  Oops
</button>
`;

exports[`Exit button component Render redirect if user is not logged in 1`] = `null`;

exports[`Exit button component Use className prop 1`] = `
<button
  className="someClass"
  onClick={[Function]}
  type="button"
>
  Oops
</button>
`;
